internal/driver/custommetadata_test.go

```diff
@@ -0,0 +1,277 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package driver
+
+import (
+	"testing"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type testCase struct {
+	device           contract.Device
+	data             string
+	expectedResponse contract.ProtocolProperties
+}
+
+func TestOnvifClient_getCustomMetadata(t *testing.T) {
+
+	myDevice := contract.Device{
+		Protocols: map[string]contract.ProtocolProperties{
+			CustomMetadata: {
+				"Location":          "Front door",
+				"CommonName":        "Front door camera",
+				"Installation date": "01/01/2022",
+				"Maintenance date":  "05/01/2022",
+			},
+		},
+	}
+
+	singleDataTest := testCase{
+		data: `{
+		"CustomMetadata":[
+			"CommonName"
+		]
+		}`,
+		expectedResponse: contract.ProtocolProperties{
+			"CommonName": myDevice.Protocols[CustomMetadata]["CommonName"],
+		},
+	}
+
+	happyMultipleDataTest := testCase{
+		data: `{
+		"CustomMetadata":[
+			"Location",
+			"CommonName"
+		]
+		}`,
+		expectedResponse: contract.ProtocolProperties{
+			"Location":   myDevice.Protocols[CustomMetadata]["Location"],
+			"CommonName": myDevice.Protocols[CustomMetadata]["CommonName"],
+		},
+	}
+
+	noFieldDataTest := testCase{
+		data: `{
+		"CustomMetadata":["Movie"]
+		}`,
+		expectedResponse: contract.ProtocolProperties{
+			"Movie": "This field does not exist in custom metadata",
+		},
+	}
+
+	noFieldsDataTest := testCase{
+		data: `{
+		"CustomMetadata":["Movie", "Height"]
+		}`,
+		expectedResponse: contract.ProtocolProperties{
+			"Movie":  "This field does not exist in custom metadata",
+			"Height": "This field does not exist in custom metadata",
+		},
+	}
+	emptyData := testCase{
+		data:             `{"CustomMetadata":[]}`,
+		expectedResponse: contract.ProtocolProperties{},
+	}
+
+	tests := []struct {
+		name          string
+		device        contract.Device
+		data          string
+		expected      contract.ProtocolProperties
+		errorExpected bool
+	}{
+		{"happy path without data - getAll", myDevice, "", myDevice.Protocols[CustomMetadata], false},
+		{"happy path with data for single field", myDevice, singleDataTest.data, singleDataTest.expectedResponse, false},
+		{"happy path with data for multiple field", myDevice, happyMultipleDataTest.data, happyMultipleDataTest.expectedResponse, false},
+		{"happy path with data for non-existent field", myDevice, noFieldDataTest.data, noFieldDataTest.expectedResponse, false},
+		{"happy path with data for multiple non-existent fields", myDevice, noFieldsDataTest.data, noFieldsDataTest.expectedResponse, false},
+		{"with empty data", myDevice, emptyData.data, emptyData.expectedResponse, true},
+		{"badJson", myDevice, "bogus", myDevice.Protocols[CustomMetadata], true},
+	}
```

Comment:
For all of these test cases, you should refactor them such that there is only one test struct creation and the individual test cases are not created above using variables, but instead as inlined values as part of the tests. 

So your `tests := []struct {` and your `type testCase struct` structs should be merged into 1 struct, with all the tests being defined inline in one place.